File: lightroom_tagger/core/analyzer/scoring.py
"""Per-perspective scoring vision operation."""

from __future__ import annotations

import contextlib
import functools
import os
from typing import Any

from lightroom_tagger.core.provider_registry import ProviderRegistry
from lightroom_tagger.core.structured_output import parse_score_response_with_retry
from lightroom_tagger.core.vision_client import (
    complete_chat_text,
    generate_description,
    make_score_json_llm_fixer,
)

from .image_prep import compress_image, get_viewable_path_managed


def parse_score_vision_response(
    raw: str,
    provider: str,
    model: str,
    *,
    registry: ProviderRegistry | None = None,
    log_callback=None,
) -> tuple[Any, bool]:
    """Parse score model output with optional LLM JSON repair."""
    reg = registry or ProviderRegistry()
    client = reg.get_client(provider)

    def _log_repair(msg: str) -> None:
        if log_callback is not None:
            log_callback("info", msg)

    llm_fixer = make_score_json_llm_fixer(
        functools.partial(complete_chat_text, client, model),
    )
    return parse_score_response_with_retry(
        raw,
        llm_fixer=llm_fixer,
        log_repair=_log_repair,
    )


def build_score_op_spec(
    path: str,
    *,
    user_prompt: str,
    provider_id: str | None = None,
    model: str | None = None,
    log_callback=None,
    silent_compression: bool = False,
    registry: ProviderRegistry | None = None,
):
    """Build a :class:`VisionOpSpec` for the scoring vision operation."""
    from lightroom_tagger.core.vision_op import VisionOpSpec

    temp_files: list[str] = []

    def prepare_fn_factory():
        viewable, viewable_is_temp = get_viewable_path_managed(path)
        if viewable_is_temp:
            temp_files.append(viewable)

        if silent_compression:
            compressed = viewable
        else:
            compressed = compress_image(viewable)
            if compressed != viewable:
                temp_files.append(compressed)

        def fn_factory(client, mdl):
            return lambda: generate_description(
                client,
                mdl,
                compressed,
                log_callback=log_callback,
                user_prompt=user_prompt,
            )

        return fn_factory

    def cleanup():
        for f in temp_files:
            if os.path.exists(f):
                with contextlib.suppress(Exception):
                    os.unlink(f)

    def parse_response(raw: str, provider: str, model: str):
        return parse_score_vision_response(
            raw,
            provider,
            model,
            registry=registry,
            log_callback=log_callback,
        )

    return VisionOpSpec(
        resolve_kind="description",
        operation="score",
        provider_id=provider_id,
        model=model,
        fn_factory=prepare_fn_factory,
        parse_response=parse_response,
        log_callback=log_callback,
        registry=registry,
        _cleanup=cleanup,
    )

File: apps/visualizer/backend/src/analyzer/scoring.ts
/**
 * The per-perspective scoring vision operation.
 *
 * Same image preparation and cleanup as `buildDescriptionOpSpec`. The parser
 * validates score output and allows one LLM JSON repair attempt.
 */
import { unlink } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { compressImage, getViewablePathManaged } from '../imaging/image-prep.js';
import { ProviderRegistry } from '../providers/registry.js';
import type { CancelCheck, LogCallback } from '../providers/retry.js';
import { generateDescription, makeScoreJsonLlmFixer } from '../providers/vision-client.js';
import {
  parseScoreResponseWithRetry,
  type ScoreResponse,
} from '../vision/structured-output.js';
import type { VisionOpSpec } from '../vision/vision-op.js';

/** A validated score, plus whether it survived only because repair rescued it. */
export interface ParsedScore {
  parsed: ScoreResponse;
  repaired: boolean;
}

/** Validate score output, with one LLM JSON repair attempt against the same model. */
export async function parseScoreVisionResponse(
  raw: string,
  provider: string,
  model: string,
  opts: { registry?: ProviderRegistry | null; logCallback?: LogCallback } = {},
): Promise<ParsedScore> {
  const registry = opts.registry ?? new ProviderRegistry();
  const client = registry.getClient(provider);

  return parseScoreResponseWithRetry(raw, {
    llmFixer: makeScoreJsonLlmFixer(client, model),
    logRepair: (message) => opts.logCallback?.('info', message),
  });
}

export interface ScoreOpOptions {
  userPrompt: string;
  providerId?: string | null;
  model?: string | null;
  logCallback?: LogCallback;
  /** The path is an already-compressed cache file; do not recompress it. */
  silentCompression?: boolean;
  registry?: ProviderRegistry | null;
  think?: boolean;
  maxTokens?: number;
  cancelCheck?: CancelCheck;
}

/**
 * Build the spec for the scoring vision op.
 *
 * `resolveKind` is `'description'`, not `'score'`: scoring runs on the same
 * vision model as describe and there is no separate configured default for it.
 */
export function buildScoreOpSpec(path: string, opts: ScoreOpOptions): VisionOpSpec<ParsedScore> {
  const tempFiles: string[] = [];

  return {
    resolveKind: 'description',
    operation: 'score',
    providerId: opts.providerId ?? null,
    model: opts.model ?? null,
    logCallback: opts.logCallback ?? null,
    registry: opts.registry ?? null,
    cancelCheck: opts.cancelCheck ?? null,
    parseResponse: (raw, provider, model) =>
      parseScoreVisionResponse(raw, provider, model, {
        registry: opts.registry ?? null,
        logCallback: opts.logCallback ?? null,
      }),
    fnFactory: async () => {
      const viewable = await getViewablePathManaged(path);
      if (viewable.isTemp) tempFiles.push(viewable.path);

      let compressed = viewable.path;
      if (!opts.silentCompression) {
        compressed = await compressImage(viewable.path);
        if (compressed !== viewable.path) tempFiles.push(compressed);
      }

      return (client, mdl) => async () =>
        generateDescription(client, mdl, compressed, {
          logCallback: opts.logCallback ?? null,
          userPrompt: opts.userPrompt,
          think: opts.think ?? false,
          maxTokens: opts.maxTokens ?? 2048,
          // Unreachable: `buildScoringUserPrompt` never returns empty. Naming the
          // describe prompt here instead would ask the model for the wrong schema.
          fallbackPrompt: opts.userPrompt,
        });
    },
    cleanup: async () => {
      for (const f of tempFiles) {
        if (existsSync(f)) await unlink(f).catch(() => undefined);
      }
    },
  };
}
